### DISCUSSION TITLE
[Unity] if you exit the program, the port will not close and will be occupied.

### DISCUSSION THREAD

**Discussion by 1101728133** [alec1o/Netly#29] [Q&A (Question and Answer)] [answered]
You need to add ```private void OnApplicationQuit() => server.Close();``` Can Netly automatically handle this issue

**Comment by alec1o** [comment, marked as answer]
##### Can Netly automatically handle this issue?
<sub>No! ``OnApplicationQuit() => server.Close();`` is only available to ``MonoBehaviour`` children. The another way is using ``Application.quitting += server.Close()``</sub>

> The unique good way is editing source code: ``Netly/src/Udp/UdpServer.cs``
##### UdpServer.cs
```cs
namespace Netly
{
    public class UdpServer
    {
        public UdpServer(bool useConnection, int timeout = 10000)
        {
#if UNITY_ENGINE
            UnityEngine.Application.quitting += this.Close();
#endif
        }
    }
}
```

##### Netly Code. ``Easy way``
```cs
using System;
using Netly;
using Netly.Core;

public class Example: MonoBehaviour
{
    private UdpServer server = new UdpServer();

    private void OnApplicationQuit()
    {
        server?.Close();
    }

    private void Awake()
    {
        server.OnOpen(() => print("server connected!"));
        server.OnError((exception) => print($"server error on connecting: {exception}"));
        server.OnClose(() => print("server disconnected!"));
        server.OnData((client, buffer) => print($"server received data from {client.Host} -> {NE.GetString(buffer)}"));
    }

    private void Start()
    {
        server.Open(new Host("127.0.0.1", 8080));
    }
}
```

### Notice:
This only happens in the unity editor and not in publication. because when a program is closed all unity threads will be destroyed by the operating system.

### Why netly netly require closed on ``OnApplicationQuit`` (Unity)?
In short, unity doesn't completely destroy threads in editor mode, this means Netly remains active even outside of game mode. Unity only completely destroys passed threads when it starts playmode again or unity editor is closed.

### And why doesn't it self-destruct when the app is closed?
Netly is automatically destroyed, but in the Unity Editor this is not the case because netly is a library that is not exclusive to Unity, it is not aware of Unity's functions and does not even know that Unity will be terminated.


### Explain.
Netly does not have any "direct connection" to unity. That is: Netly does not call or know any unity function because netly is an independent library and is not exclusive to unity, therefore netly does not have the knowledge of unity methods, and the reason why the connection is not well closed and Because unity does not completely destroy all background processes created by the runtime in the unity editor version, the thread to netly continues running since netly does not have any "direct connection'' with unity.

And the only point that connects netly and unity and "MainThread" "(Plugins/Netly/Wrapper/NETLY_THREAD.cs)" this script makes netly dispatch messages in the unity thread (because unity blocks external calls with threads extrageira: it mean you wouldn't create a GameObject, Change the label of a Text, among others on netly callbacks) so you need to manually close the connection because netly never knows if it runs on the console, on unity, on flax, .. .. I hope you understand and I believe it was a comprehensible explanation due to this reason I had asked to open a new discussion.
